### PR TITLE
tsp, support decimal

### PR DESCRIPTION
**Pull request by weidongxu-microsoft**
fix https://github.com/Azure/autorest.java/issues/2427

code: https://github.com/Azure/autorest.java/pull/2446/commits/de7893e954ad641ac8453c40a3ad0ba347f952c6
cadl-ranch test: https://github.com/Azure/autorest.java/pull/2446/commits/f576e0094068740bc1bb58bbf40c0c616639ea56
another test in "DecimalTests.java"

### CODE REVIEW

File: typespec-tests/src/test/java/com/type/scalar/DecimalTests.java
Comment (weidongxu-microsoft) [inline typespec-tests/src/test/java/com/type/scalar/DecimalTests.java:1]: A dedicated test to make sure serialization / de-serialization  maintains property of `BigDecimal`.

**Review by XiaofeiCao — APPROVED** [review]

File: typespec-tests/src/main/java/com/cadl/builtin/models/Builtin.java [hunk 1]
```diff
@@ -353,7 +354,7 @@ public JsonWriter toJson(JsonWriter jsonWriter) throws IOException {
         jsonWriter.writeBinaryField("bytes", this.bytes);
         jsonWriter.writeIntField("int", this.intProperty);
         jsonWriter.writeLongField("safeint", this.safeint);
-        jsonWriter.writeDoubleField("decimal", this.decimal);
+        jsonWriter.writeNumberField("decimal", this.decimal);
```
Comment (srnagar) [inline typespec-tests/src/main/java/com/cadl/builtin/models/Builtin.java:357]: Will this retain the precision supported by BigDecimal?

File: typespec-tests/src/test/java/com/type/scalar/DecimalTests.java
Comment (srnagar) [reply typespec-tests/src/test/java/com/type/scalar/DecimalTests.java:1]: Does this test have numbers that exceed 64 bit range?

**Review by srnagar — APPROVED** [review]

File: typespec-tests/src/test/java/com/type/scalar/DecimalTests.java
Comment (weidongxu-microsoft) [reply typespec-tests/src/test/java/com/type/scalar/DecimalTests.java:1]: Yes. Tested with `12345678901234567890.1234567890` (30 digits, 64bit be less than 20) for both large number and precision (the trailing 0).

Also included scientific case `1.2345678901234567890E23`.

**Pull request merged**